### PR TITLE
[ALLUXIO-1911] Separate initialization path for Configuration

### PR DESCRIPTION
**Pull request by apc999**
https://alluxio.atlassian.net/browse/ALLUXIO-1911

### CODE REVIEW

File: core/common/src/main/java/alluxio/Configuration.java [hunk 1]
```diff
@@ -167,7 +157,7 @@ public Configuration() {
     mProperties.putAll(siteProps);
     mProperties.putAll(systemProps);
 
-    // Update alluxio.master_address based on if Zookeeper is used or not.
+    // Update alluxio.master.address based on if Zookeeper is used or not.
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/Configuration.java]: I think this comment should be removed completely, it is outdated.

File: core/common/src/main/java/alluxio/RuntimeConstants.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
+/*
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/RuntimeConstants.java]: would it make sense to just make this part of the `Constants.java` file? 


the only place where `Constants.VERSION` is used is here so we could just not expose it (or adopt a convention that appends `_KEY` to all the property keys ... which I think would be useful anyhow)

File: core/common/src/main/java/alluxio/Configuration.java [hunk 3]
```diff
@@ -104,11 +93,9 @@ public Configuration(Properties props) {
 
   /**
    * Default constructor.
-   *
-   * Most clients will call this constructor to allow default loading of properties to happen.
    */
   public Configuration() {
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/Configuration.java]: should this be made private to force uses to use the factory methods?

File: core/common/src/main/java/alluxio/Configuration.java [hunk 4]
```diff
@@ -70,32 +71,20 @@
...
-    checkUserFileBufferBytes();
-  }
-
-  /**
-   * Overrides default properties.
-   *
-   * @param props override {@link Properties}
-   */
-  public Configuration(Properties props) {
+  public Configuration(Map<?, ?> props) {
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/Configuration.java]: should this be made private to force uses to use the factory methods?

**Comment by jsimsa** [comment]
@apc999 thanks for the PR, did a pass and left some comments

**Comment by jsimsa** [comment]
@apc999 thanks for the PR, did a pass and left some comments, also please fix the build failure

File: core/common/src/main/java/alluxio/RuntimeConstants.java (hunk 2 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/RuntimeConstants.java]: it was in `Constants.java` before, but causing the problem of introducing circular dependency

Comment (apc999) [inline core/common/src/main/java/alluxio/RuntimeConstants.java]: The difference (I want to set after this PR) is, `Constants.java` serving constants determined in compile time, and `RuntimeConstants` serving constants determined once the process is launched.

Comment (jsimsa) [inline core/common/src/main/java/alluxio/RuntimeConstants.java]: what was the circular dependency and how does it manifest in Jenkins?

would making Runtime a nested class of Constants work?

File: core/common/src/main/java/alluxio/RuntimeConstants.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    VERSION = configuration.get(Constants.VERSION);
+  }
+
+  /** The version of this Alluxio instance */
+  public static final String VERSION;
+
+  /** The relative path to the Alluxio target jar. */
+  public static final String ALLUXIO_JAR =
+      "target/alluxio-" + VERSION + "-jar-with-dependencies.jar";
+
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/RuntimeConstants.java]: add

```
private RuntimeConstants() // prevent instantiation
```

File: core/common/src/main/java/alluxio/RuntimeConstants.java (hunk 2 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/RuntimeConstants.java]: @aaudiber has more context on the bug (https://github.com/Alluxio/alluxio/pull/1905). basically, if code here moved to `Constants.java`, the static block of `Constants`and `Configuration` depends on each other, causing unstable build problem. Moving `RuntimeConstant` nested under `Constant` will not solve that problem

Comment (aaudiber) [inline core/common/src/main/java/alluxio/RuntimeConstants.java]: The constructor of `Configuration` (called in the static block here) makes use of many fields of `Constants`. Moving this static initialization block to `Constants` will make it impossible to statically initialize `Constants` without constructing a `Configuration`, and impossible to construct a `Configuration` without having `Constants` be statically initialized. To avoid this problem, we should make sure that `Constants` can be statically initialized without depending on any other classes.

Comment (jsimsa) [inline core/common/src/main/java/alluxio/RuntimeConstants.java]: thanks for the explanation ... on a related note, do you guys think it would make sense to move all the constants that are property key names to a separate class (e.g. PropertyKey), I find it mildly confusing that Constants contains both property key names and various other constants. I think `conf.get(PropertyKey.VERSION)` reads better than `conf.get(Constants.VERSION)` this also avoids confusion between `Constants.VERSION` and `RuntimeConstants.VERSION`

Comment (apc999) [inline core/common/src/main/java/alluxio/RuntimeConstants.java]: @jsimsa I support this idea. And I think it could be a separate enum Type to be more type safe.

Comment (jsimsa) [inline core/common/src/main/java/alluxio/RuntimeConstants.java]: sgtm

Comment (apc999) [inline core/common/src/main/java/alluxio/RuntimeConstants.java]: but that probably should be in another PR. (A JIRA added ALLUXIO-1949).

Comment (jsimsa) [inline core/common/src/main/java/alluxio/RuntimeConstants.java]: makes sense

File: core/common/src/main/java/alluxio/Configuration.java (hunk 3 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/Configuration.java]: agree. Let's see how far we can go in this PR. Too many references of this constructor (especially in test), and I don't want to further enlarge the footprint of this PR.

Comment (jsimsa) [inline core/common/src/main/java/alluxio/Configuration.java]: I think there is actually value of replacing all the references in this PR ... that way we know that the default / worker / master / client configuration is used where it should be used

File: core/common/src/main/java/alluxio/Configuration.java (hunk 4 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/Configuration.java]: same as above, this should be the goal and let's see how far we go in this PR (not having a too big PR)

File: core/common/src/main/java/alluxio/Configuration.java (hunk 3 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/Configuration.java]: I removed the references to this `public constructor` in production code in this PR.
for test code, there are > 60 references, I think it might be better to have separate PR doing that.

File: core/common/src/main/java/alluxio/RuntimeConstants.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  /** The version of this Alluxio instance. */
+  public static final String VERSION;
+
+  /** The relative path to the Alluxio target jar. */
+  public static final String ALLUXIO_JAR =
+      "target/alluxio-" + VERSION + "-jar-with-dependencies.jar";
+
+  private RuntimeConstants() {
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/RuntimeConstants.java]: can this be just one line?

**Comment by jsimsa** [comment]
one minor nit, once the failing tests are fix this LGTM

File: core/common/src/main/java/alluxio/RuntimeConstants.java (hunk 6 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/RuntimeConstants.java]: done

File: core/common/src/main/java/alluxio/Configuration.java [hunk 7]
```diff
@@ -518,4 +497,49 @@ private void checkUserFileBufferBytes() {
...
+     */
+    public static Configuration create(Map<?, ?> properties) {
+      return new Configuration(properties);
+    }
+
+    /**
+     * @param conf the source configuration
+     * @return a new configuration with only properties copied from source
+     */
+    public static Configuration create(Configuration conf) {
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/Configuration.java]: should this be called clone?

File: core/common/src/main/java/alluxio/Configuration.java [hunk 8]
```diff
@@ -518,4 +497,49 @@ private void checkUserFileBufferBytes() {
...
+   * Factory class to create configuration instances.
+   */
+  @ThreadSafe
+  public static final class Factory {
+
+    /**
+     * @param properties the source configuration
+     * @return a new configuration with only properties copied from source
+     */
+    public static Configuration create(Map<?, ?> properties) {
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/Configuration.java]: should this be called `fromProperties`?

File: examples/src/main/java/alluxio/examples/BasicNonByteBufferOperations.java [hunk 9]
```diff
@@ -123,4 +123,5 @@ private boolean read(FileSystem alluxioClient) throws IOException, AlluxioExcept
     }
     return true;
   }
+
```
Comment (jsimsa) [inline examples/src/main/java/alluxio/examples/BasicNonByteBufferOperations.java]: nit: not needed

**Comment by jsimsa** [comment]
@apc999 thanks, did another pass and left some minor comments

File: core/common/src/main/java/alluxio/Configuration.java (hunk 7 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/Configuration.java]: updated.

File: core/common/src/main/java/alluxio/Configuration.java (hunk 8 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/Configuration.java]: done. And also moved from Factory class to object class---I think it is less ideal to have Factory here since we are not really exposing an interface.

File: examples/src/main/java/alluxio/examples/BasicNonByteBufferOperations.java (hunk 9 above)
Comment (apc999) [inline examples/src/main/java/alluxio/examples/BasicNonByteBufferOperations.java]: fixed

**Comment by jsimsa** [comment]
LGTM

File: core/common/src/main/java/alluxio/Configuration.java (hunk 7 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/Configuration.java]: after doing some search, I think implementing `clone` is not considered  a good practice compared to using copy factory method or constructor. more details can be found 
http://stackoverflow.com/questions/11905630/java-super-clone-method-and-inheritance

Comment (jsimsa) [inline core/common/src/main/java/alluxio/Configuration.java]: sgtm, so you will use a copy constructor?

File: core/common/src/main/java/alluxio/Configuration.java [hunk 10]
```diff
@@ -75,10 +75,18 @@
   private final Properties mProperties = new Properties();
 
   /**
+   * @param conf the source configuration to copy
+   * @return a new configuration instance with the properties copied from source configuration
+   */
+  public static Configuration copyFrom(Configuration conf) {
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/Configuration.java]: OOC, why not use a copy constructor?

File: core/common/src/main/java/alluxio/Configuration.java [hunk 11]
```diff
@@ -55,47 +54,69 @@
...
- * <p>
- * The class only supports creation using {@link #Configuration(Properties)} to override default
- * values.
  */
 @NotThreadSafe
 public final class Configuration {
   /** File to set default properties. */
   public static final String DEFAULT_PROPERTIES = "alluxio-default.properties";
-  /** File to set customized properties. */
+  /** File to set customized properties for Alluxio deployment. */
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/Configuration.java]: Does this mean both client and server? If so I think its clearer to just state it explicitly.

File: core/server/src/main/java/alluxio/cli/ValidateConf.java [hunk 12]
```diff
@@ -36,9 +36,10 @@
   /**
    * Validates the configurations.
    *
+   * @param configuration Configuration to validate
    * @return true if the validation succeeds, false otherwise
    */
-  public static boolean validate() {
+  public static boolean validate(Configuration configuration) {
```
Comment (calvinjia) [inline core/server/src/main/java/alluxio/cli/ValidateConf.java]: Should we make it clear that the command now only validates the client configuration?

File: core/server/src/main/java/alluxio/master/AlluxioMaster.java [hunk 13]
```diff
@@ -71,12 +71,13 @@
...
     if (args.length != 0) {
-      LOG.info("java -cp {} {}", Version.ALLUXIO_JAR, AlluxioMaster.class.getCanonicalName());
+      LOG.info("java -cp {} {}", RuntimeConstants.ALLUXIO_JAR,
+          AlluxioMaster.class.getCanonicalName());
       System.exit(-1);
     }
 
     // validate the conf
-    if (!ValidateConf.validate()) {
+    if (!ValidateConf.validate(Configuration.createServerConf())) {
```
Comment (calvinjia) [inline core/server/src/main/java/alluxio/master/AlluxioMaster.java]: Should we validate the client conf whenever we start up the client context as well?

File: core/common/src/main/java/alluxio/Configuration.java [hunk 14]
```diff
@@ -55,47 +54,69 @@
...
+   * @return a new configuration with only properties copied from source
+   */
+  public static Configuration fromMap(Map<?, ?> properties) {
+    return new Configuration(properties);
+  }
+
+  /**
+   * @return the default configuration without loading site and system properties
+   */
+  public static Configuration createDefaultConf() {
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/Configuration.java]: Could you add a TODO for unit testing the new constructors?

**Comment by calvinjia** [comment]
@apc999 left a few comments, looks good overall.

File: core/common/src/main/java/alluxio/Configuration.java (hunk 10 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/Configuration.java]: yeah, I think copy constructor is better. Changed.

File: core/common/src/main/java/alluxio/Configuration.java (hunk 11 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/Configuration.java]: done

File: core/server/src/main/java/alluxio/cli/ValidateConf.java (hunk 12 above)
Comment (apc999) [inline core/server/src/main/java/alluxio/cli/ValidateConf.java]: not really. this method is also invoked by the master and worker. On the main, it is doing client-side check.

File: core/server/src/main/java/alluxio/master/AlluxioMaster.java (hunk 13 above)
Comment (apc999) [inline core/server/src/main/java/alluxio/master/AlluxioMaster.java]: I am concerned by dong such a change: because that may bring down client-side job, right?

File: core/common/src/main/java/alluxio/Configuration.java (hunk 14 above)
Comment (apc999) [inline core/common/src/main/java/alluxio/Configuration.java]: that's in the followup PR (adding unit tests for those constructors and utils)
TODO added

**Comment by apc999** [comment]
@calvinjia  PTAL thanks

File: core/server/src/main/java/alluxio/master/AlluxioMaster.java (hunk 13 above)
Comment (calvinjia) [inline core/server/src/main/java/alluxio/master/AlluxioMaster.java]: Yes, it could cause clients to exit unexpectedly, but that might be the correct behavior if the configuration is flawed? Otherwise the client behavior might be unintended.

File: core/server/src/main/java/alluxio/cli/ValidateConf.java [hunk 15]
```diff
@@ -103,10 +103,19 @@ public static boolean validate() {
...
+    }
+
+    System.out.println("Validating client configuration.");
+    if (validate(Configuration.createServerConf())) {
+      System.out.println("All server configuration entries are valid.");
+    } else {
+      System.out.println("Server configuration has invalid entries.");
     }
-    System.out.println("All configuration entries are valid.");
     System.exit(0);
```
Comment (calvinjia) [inline core/server/src/main/java/alluxio/cli/ValidateConf.java]: Should we `exit(-1)` still if either of the configurations are invalid?

**Comment by calvinjia** [comment]
@apc999 Looks good overall, one minor comment. For validating the client configuration, I don't feel strongly about it if you think it will cause more trouble than benefit, but since the configuration change will be nontrivial, it may be worth making sure users don't have invalid configurations.

File: core/server/src/main/java/alluxio/master/AlluxioMaster.java (hunk 13 above)
Comment (apc999) [inline core/server/src/main/java/alluxio/master/AlluxioMaster.java]: I updated the PR so alluxio shell will check the client-side conf before running.
for client conf in general, I didn't check it we don't have a good entrance point to check it (as in master or worker or shell, we have the main to check). It is possible that we could have it in `createClientConf`, but then it will be strange why we check in `createClientConf` but not in `createServerConf`

**Comment by apc999** [comment]
@calvinjia I updated the PR a bunch to clean up the configuration check . especially, I added check to alluxio shell too. I explained the reason it is not trivial to check client side in general. We could discuss it for sure.

**Comment by calvinjia** [comment]
LGTM

**Comment by apc999** [comment]
thanks for the review. I will submit followup PRs for unit tests later.
